### PR TITLE
tests: refactored fs watch_dir tests for stability

### LINKED ISSUE #30: test: fs_event_watch_dir fails on OSX
Thank you for your awesome work guys. :+1: 

Building @ Darwin, got this error :

````
deps/libuv [ sh autogen.sh;./configure;make;make check;make install                                                                    master ] 2:54 pm
+ glibtoolize
glibtoolize: putting auxiliary files in `.'.
glibtoolize: linking file `./ltmain.sh'
glibtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
glibtoolize: linking file `m4/libtool.m4'
glibtoolize: linking file `m4/ltoptions.m4'
glibtoolize: linking file `m4/ltsugar.m4'
glibtoolize: linking file `m4/ltversion.m4'
glibtoolize: linking file `m4/lt~obsolete.m4'
+ aclocal -I m4
+ autoconf
+ automake --add-missing --copy
configure.ac:29: installing './ar-lib'
configure.ac:25: installing './compile'
configure.ac:22: installing './config.guess'
configure.ac:22: installing './config.sub'
configure.ac:21: installing './install-sh'
configure.ac:21: installing './missing'
Makefile.am: installing './depcomp'
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-apple-darwin13.4.0
checking host system type... x86_64-apple-darwin13.4.0
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking if gcc supports -Wno-dollar-in-identifier-extension flag... yes
checking for ar... ar
checking the archiver (ar) interface... ar
checking whether make supports nested variables... (cached) yes
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld
checking if the linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm
checking the name lister (/usr/bin/nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 196608
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-apple-darwin13.4.0 file names to x86_64-apple-darwin13.4.0 format... func_convert_file_noop
checking how to convert x86_64-apple-darwin13.4.0 file names to toolchain format... func_convert_file_noop
checking for /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for archiver @FILE support... no
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dsymutil... dsymutil
checking for nmedit... nmedit
checking for lipo... lipo
checking for otool... otool
checking for otool64... no
checking for -single_module linker flag... yes
checking for -exported_symbols_list linker flag... yes
checking for -force_load linker flag... yes
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... yes
checking for gcc option to produce PIC... -fno-common -DPIC
checking if gcc PIC flag -fno-common -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin13.4.0 dyld
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for dlopen in -ldl... yes
checking for kstat_lookup in -lkstat... no
checking for kvm_open in -lkvm... no
checking for gethostbyname in -lnsl... no
checking for perfstat_cpu in -lperfstat... no
checking for pthread_mutex_init in -lpthread... yes
checking for clock_gettime in -lrt... no
checking for sendfile in -lsendfile... no
checking for socket in -lsocket... no
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for pkg-config... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating libuv.pc
config.status: creating Makefile
config.status: executing depfiles commands
config.status: executing libtool commands
  CC       src/libuv_la-fs-poll.lo
  CC       src/libuv_la-inet.lo
  CC       src/libuv_la-threadpool.lo
  CC       src/libuv_la-uv-common.lo
  CC       src/libuv_la-version.lo
  CC       src/unix/libuv_la-async.lo
  CC       src/unix/libuv_la-core.lo
  CC       src/unix/libuv_la-dl.lo
  CC       src/unix/libuv_la-fs.lo
  CC       src/unix/libuv_la-getaddrinfo.lo
  CC       src/unix/libuv_la-getnameinfo.lo
  CC       src/unix/libuv_la-loop-watcher.lo
  CC       src/unix/libuv_la-loop.lo
  CC       src/unix/libuv_la-pipe.lo
  CC       src/unix/libuv_la-poll.lo
  CC       src/unix/libuv_la-process.lo
  CC       src/unix/libuv_la-signal.lo
  CC       src/unix/libuv_la-stream.lo
src/unix/stream.c:1054:19: warning: variable length array folded to constant array as an extension [-Wgnu-folding-constant]
  char cmsg_space[CMSG_SPACE(UV__CMSG_FD_SIZE)];
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/socket.h:517:24: note: expanded from macro 'CMSG_SPACE'
#define CMSG_SPACE(l)           (__DARWIN_ALIGN32(sizeof(struct cmsghdr)) + __DARWIN_ALIGN32(l))
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/unix/stream.c:1054:19: note: cast that performs the conversions of a reinterpret_cast is not allowed in a constant expression
/usr/include/sys/socket.h:517:25: note: expanded from macro 'CMSG_SPACE'
#define CMSG_SPACE(l)           (__DARWIN_ALIGN32(sizeof(struct cmsghdr)) + __DARWIN_ALIGN32(l))
                                 ^
/usr/include/i386/_param.h:43:42: note: expanded from macro '__DARWIN_ALIGN32'
#define       __DARWIN_ALIGN32(p)       ((__darwin_size_t)((char *)(__darwin_size_t)(p) + __DARWIN_ALIGNBYTES32) &~ __DARWIN_ALIGNBYTES32)
                                         ^
1 warning generated.
  CC       src/unix/libuv_la-tcp.lo
  CC       src/unix/libuv_la-thread.lo
  CC       src/unix/libuv_la-timer.lo
  CC       src/unix/libuv_la-tty.lo
  CC       src/unix/libuv_la-udp.lo
  CC       src/unix/libuv_la-darwin.lo
  CC       src/unix/libuv_la-darwin-proctitle.lo
  CC       src/unix/libuv_la-fsevents.lo
  CC       src/unix/libuv_la-kqueue.lo
  CC       src/unix/libuv_la-proctitle.lo
  CCLD     libuv.la
/Applications/Xcode.app/Contents/Developer/usr/bin/make  test/run-tests
  CC       test/test_run_tests-blackhole-server.o
  CC       test/test_run_tests-dns-server.o
  CC       test/test_run_tests-echo-server.o
  CC       test/test_run_tests-run-tests.o
  CC       test/test_run_tests-runner.o
  CC       test/test_run_tests-test-active.o
  CC       test/test_run_tests-test-async.o
  CC       test/test_run_tests-test-async-null-cb.o
  CC       test/test_run_tests-test-barrier.o
  CC       test/test_run_tests-test-callback-order.o
  CC       test/test_run_tests-test-callback-stack.o
  CC       test/test_run_tests-test-close-fd.o
  CC       test/test_run_tests-test-close-order.o
  CC       test/test_run_tests-test-condvar.o
  CC       test/test_run_tests-test-connection-fail.o
  CC       test/test_run_tests-test-cwd-and-chdir.o
  CC       test/test_run_tests-test-default-loop-close.o
  CC       test/test_run_tests-test-delayed-accept.o
  CC       test/test_run_tests-test-dlerror.o
  CC       test/test_run_tests-test-embed.o
  CC       test/test_run_tests-test-emfile.o
  CC       test/test_run_tests-test-error.o
  CC       test/test_run_tests-test-fail-always.o
  CC       test/test_run_tests-test-fs-event.o
  CC       test/test_run_tests-test-fs-poll.o
  CC       test/test_run_tests-test-fs.o
  CC       test/test_run_tests-test-get-currentexe.o
  CC       test/test_run_tests-test-get-loadavg.o
  CC       test/test_run_tests-test-get-memory.o
  CC       test/test_run_tests-test-getaddrinfo.o
  CC       test/test_run_tests-test-getnameinfo.o
  CC       test/test_run_tests-test-getsockname.o
  CC       test/test_run_tests-test-handle-fileno.o
  CC       test/test_run_tests-test-hrtime.o
  CC       test/test_run_tests-test-idle.o
  CC       test/test_run_tests-test-ip4-addr.o
  CC       test/test_run_tests-test-ip6-addr.o
  CC       test/test_run_tests-test-ipc-send-recv.o
  CC       test/test_run_tests-test-ipc.o
  CC       test/test_run_tests-test-loop-handles.o
  CC       test/test_run_tests-test-loop-alive.o
  CC       test/test_run_tests-test-loop-close.o
  CC       test/test_run_tests-test-loop-stop.o
  CC       test/test_run_tests-test-loop-time.o
  CC       test/test_run_tests-test-multiple-listen.o
  CC       test/test_run_tests-test-mutexes.o
  CC       test/test_run_tests-test-osx-select.o
  CC       test/test_run_tests-test-pass-always.o
  CC       test/test_run_tests-test-ping-pong.o
  CC       test/test_run_tests-test-pipe-bind-error.o
  CC       test/test_run_tests-test-pipe-connect-error.o
  CC       test/test_run_tests-test-pipe-getsockname.o
  CC       test/test_run_tests-test-pipe-sendmsg.o
  CC       test/test_run_tests-test-pipe-server-close.o
  CC       test/test_run_tests-test-pipe-close-stdout-read-stdin.o
  CC       test/test_run_tests-test-platform-output.o
  CC       test/test_run_tests-test-poll-close.o
  CC       test/test_run_tests-test-poll-closesocket.o
  CC       test/test_run_tests-test-poll.o
  CC       test/test_run_tests-test-process-title.o
  CC       test/test_run_tests-test-ref.o
  CC       test/test_run_tests-test-run-nowait.o
  CC       test/test_run_tests-test-run-once.o
  CC       test/test_run_tests-test-semaphore.o
  CC       test/test_run_tests-test-shutdown-close.o
  CC       test/test_run_tests-test-shutdown-eof.o
  CC       test/test_run_tests-test-shutdown-twice.o
  CC       test/test_run_tests-test-signal-multiple-loops.o
  CC       test/test_run_tests-test-signal.o
  CC       test/test_run_tests-test-socket-buffer-size.o
  CC       test/test_run_tests-test-spawn.o
  CC       test/test_run_tests-test-stdio-over-pipes.o
  CC       test/test_run_tests-test-tcp-bind-error.o
  CC       test/test_run_tests-test-tcp-bind6-error.o
  CC       test/test_run_tests-test-tcp-close-accept.o
  CC       test/test_run_tests-test-tcp-close-while-connecting.o
  CC       test/test_run_tests-test-tcp-close.o
  CC       test/test_run_tests-test-tcp-connect-error-after-write.o
  CC       test/test_run_tests-test-tcp-connect-error.o
  CC       test/test_run_tests-test-tcp-connect-timeout.o
  CC       test/test_run_tests-test-tcp-connect6-error.o
  CC       test/test_run_tests-test-tcp-flags.o
  CC       test/test_run_tests-test-tcp-open.o
  CC       test/test_run_tests-test-tcp-read-stop.o
  CC       test/test_run_tests-test-tcp-shutdown-after-write.o
  CC       test/test_run_tests-test-tcp-unexpected-read.o
  CC       test/test_run_tests-test-tcp-write-to-half-open-connection.o
  CC       test/test_run_tests-test-tcp-write-after-connect.o
  CC       test/test_run_tests-test-tcp-writealot.o
  CC       test/test_run_tests-test-tcp-try-write.o
  CC       test/test_run_tests-test-tcp-write-queue-order.o
  CC       test/test_run_tests-test-thread-equal.o
  CC       test/test_run_tests-test-thread.o
  CC       test/test_run_tests-test-threadpool-cancel.o
  CC       test/test_run_tests-test-threadpool.o
  CC       test/test_run_tests-test-timer-again.o
  CC       test/test_run_tests-test-timer-from-check.o
  CC       test/test_run_tests-test-timer.o
  CC       test/test_run_tests-test-tty.o
  CC       test/test_run_tests-test-udp-bind.o
  CC       test/test_run_tests-test-udp-dgram-too-big.o
  CC       test/test_run_tests-test-udp-ipv6.o
  CC       test/test_run_tests-test-udp-multicast-interface.o
  CC       test/test_run_tests-test-udp-multicast-interface6.o
  CC       test/test_run_tests-test-udp-multicast-join.o
  CC       test/test_run_tests-test-udp-multicast-join6.o
  CC       test/test_run_tests-test-udp-multicast-ttl.o
  CC       test/test_run_tests-test-udp-open.o
  CC       test/test_run_tests-test-udp-options.o
  CC       test/test_run_tests-test-udp-send-and-recv.o
  CC       test/test_run_tests-test-udp-send-immediate.o
  CC       test/test_run_tests-test-udp-send-unreachable.o
  CC       test/test_run_tests-test-udp-try-send.o
  CC       test/test_run_tests-test-walk-handles.o
  CC       test/test_run_tests-test-watcher-cross-stop.o
  CC       test/test_run_tests-runner-unix.o
  CCLD     test/run-tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
[%   0|+   0|-   0|T   0|S   0]: platform_output
Output from process `platform_output`:
uv_get_process_title: /Users/hems/git/loopcast/appcast/Appcast/deps/libuv/test/.libs/run-tests
uv_resident_set_memory: 593920
uv_uptime: 416572.000000
uv_getrusage:
  user: 0 sec 972 microsec
  system: 0 sec 1109 microsec
uv_cpu_info:
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 35976870
  times.user: 69336390
  times.idle: 165557690
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 6551090
  times.user: 42575950
  times.idle: 221250190
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 17615010
  times.user: 56376690
  times.idle: 196385620
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4817910
  times.user: 40160400
  times.idle: 225398880
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 15460210
  times.user: 52943520
  times.idle: 201973560
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4316560
  times.user: 39211610
  times.idle: 226849000
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 14397410
  times.user: 51247620
  times.idle: 204732230
  times.irq: 0
  times.nice: 0
  model: Intel(R) Core(TM) i7-2635QM CPU @ 2.00GHz
  speed: 2000
  times.sys: 4071700
  times.user: 38768140
  times.idle: 227537290
  times.irq: 0
  times.nice: 0
uv_interface_addresses:
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: ::1
  netmask: ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: 127.0.0.1
  netmask: 255.0.0.0
  name: lo0
  internal: 1
  physical address: 00:00:00:00:00:00
  address: fe80::1
  netmask: ffff:ffff:ffff:ffff::
  name: en1
  internal: 0
  physical address: e0:f8:47:43:8f:6c
  address: fe80::e2f8:47ff:fe43:8f6c
  netmask: ffff:ffff:ffff:ffff::
  name: en1
  internal: 0
  physical address: e0:f8:47:43:8f:6c
  address: 192.168.0.2
  netmask: 255.255.255.0
=============================================================
[%  72|+ 177|-   0|T   0|S   0]: spawn_reads_child_path
`spawn_reads_child_path` failed: exit code 6
Output from process `spawn_reads_child_path`:
exit_cb
Assertion failed in test/test-spawn.c on line 66: exit_status == 1
=============================================================
[%  77|+ 188|-   1|T   0|S   0]: osx_selectgot some input
with a couple of lines
feel pretty happy
[%  77|+ 189|-   1|T   0|S   0]: osx_select_many_fdsgot some input
with a couple of lines
feel pretty happy
[%  86|+ 211|-   1|T   0|S   0]: fs_event_watch_dir
`fs_event_watch_dir` failed: timeout
Output from process `fs_event_watch_dir`: (no output)
=============================================================
[% 100|+ 243|-   2|T   0|S   0]: Done.
FAIL: test/run-tests
======================================================
1 of 1 test failed
Please report to https://github.com/libuv/libuv/issues
======================================================
make[1]: *** [check-TESTS] Error 1
make: *** [check-am] Error 2
 ./install-sh -c -d '/usr/local/lib'
 /bin/sh ./libtool   --mode=install /usr/bin/install -c   libuv.la '/usr/local/lib'
libtool: install: /usr/bin/install -c .libs/libuv.1.dylib /usr/local/lib/libuv.1.dylib
libtool: install: (cd /usr/local/lib && { ln -s -f libuv.1.dylib libuv.dylib || { rm -f libuv.dylib && ln -s libuv.1.dylib libuv.dylib; }; })
libtool: install: /usr/bin/install -c .libs/libuv.lai /usr/local/lib/libuv.la
libtool: install: /usr/bin/install -c .libs/libuv.a /usr/local/lib/libuv.a
libtool: install: chmod 644 /usr/local/lib/libuv.a
libtool: install: ranlib /usr/local/lib/libuv.a
 ./install-sh -c -d '/usr/local/include'
 /usr/bin/install -c -m 644 include/uv.h include/uv-errno.h include/uv-threadpool.h include/uv-version.h include/uv-unix.h include/uv-darwin.h '/usr/local/include'
 ./install-sh -c -d '/usr/local/lib/pkgconfig'
 /usr/bin/install -c -m 644 libuv.pc '/usr/local/lib/pkgconfig'
deps/libuv [                                                                                                                           master ] 2:57 pm
````

### PR DESCRIPTION
**Pull request by whitlockjc**
`fs_event_watch_dir` and `fs_event_watch_dir_recursive` could fail randomly
due to the way in which the tests were written.  Originally timers were
used to create, remove and recreate the test files but this could lead
to a race condition if the timeout used to delete the test files ran
before all file creation fs events were handled.  On top of that, the
file removal timer scheduled another timer to recreate the test files
and that timer's timeout could also lead to the same condition.

The refactoring removed timers for the removal/recreation of the test
files and instead the fs event callback was updated to have the
necessary logic to drive the test file removal.  We no longer recreate
the test files since it appears to be unnecessary.

With the current PR, I can run 1000 tests for `fs_event_watch_dir` and
`fs_event_watch_dir_recursive` without a single failure, where prior to this
PR they would fail 14/100 on average.

Fixes #30

### CODE REVIEW

**Comment by whitlockjc** [comment]
One thing I forgot to mention, or maybe it wasn't clear, is that the original test appeared to do two cycles of creating files and then removing the files.  This current PR only does one cycle but updating it to do two should be trivial.  It's your call as to whether I add the extra cycle back or not.

**Comment by saghul** [comment]
I'm afk until Monday. Will take a look then! Thanks!

File: test/test-fs-event.c [hunk 1]
```diff
@@ -184,38 +144,76 @@ static const char* fs_event_get_filename_in_subdir(int i) {
...
+  fprintf(stderr, "fs_event_cb_dir_multi_file called: %i\n", fs_event_cb_called);
+  fprintf(stderr, "  created: %i\n", fs_event_created);
+  fprintf(stderr, "  removed: %i\n", fs_event_removed);
+
+  if (fs_event_cb_called == fs_event_created) {
+    /* Once we've processed all create events, delete all files */
+    fs_event_unlink_files(handle);
+  } else if (fs_event_cb_called == fs_event_created + fs_event_removed) {
+    /* Once we've processed all create and delete events, stop watching */
+    ASSERT(0 == uv_fs_event_stop(handle));
```
Comment (saghul) [inline test/test-fs-event.c]: This is not really necessary, since calling `uv_close` will trigger a stop.

**Comment by saghul** [comment]
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/130/

**Comment by saghul** [comment]
Not sure if it was failing before, but SmartOS and FreeBSD are failing:

SmartOS:
~~~~
`fs_event_watch_dir` failed: exit code 134
Output from process `fs_event_watch_dir`:
Assertion failed in test/test-fs-event.c on line 175: events == UV_RENAME
~~~~

FreeBSD:
~~~~
%  86|+ 216|-   7|T   0|S  12]: fs_event_watch_dir
`fs_event_watch_dir` failed: timeout
Output from process `fs_event_watch_dir`:
fs_event_cb_dir_multi_file called: 1
  created: 16
  removed: 0
~~~~

File: test/test-fs-event.c (hunk 1 above)
Comment (whitlockjc) [inline test/test-fs-event.c]: Makes sense.  I didn't write the original test so I didn't want to change it but I will remove it if you think it's unnecessary.

**Comment by whitlockjc** [comment]
I will remove the unnecessary stuff and test on the other environments.  I will update you.

**Comment by saghul** [comment]
Thanks! Ping me me if you need me to trigger the CI.

**Comment by whitlockjc** [comment]
I have updated the PR.  I removed my debugging output *(sorry)*, removed the unnecessary assertions @saghul mentioned and updated the assertion in the `fs_event_cb_dir_multi_file` callback so we do not have assertion failures on SmartOS.  Fixing the assertion on SmartOS does not fix the tests completely as it will timeout like it does on FreeBSD but these tests were failing prior to my PR and appear to be a deeper libuv issue than a test race condition, since that is now gone.  *(A quick debugging session shows that on FreeBSD and SmartOS that the `fs_event_cb_dir_multi_file` callback is only invoked once instead instead of once per created/removed file.)*

**Comment by saghul** [comment]
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/132/

**Comment by whitlockjc** [comment]
I can confirm I've somehow broken something.  I don't know how but it seems that way.  If I build `v1.x` on FreeBSD without my change, all 16 events are processed prior to the timeout whereas with my change, I only see 1 event processed prior to the timeout.  I'll get this sorted and will update the PR.

**Comment by whitlockjc** [comment]
I've found something.  On FreeBSD and SmartOS, the tests were failing prior to this PR.  But after looking at the CI environment, it seems something changed for `fs_event_watch_dir` because it use to pass on CentOS 7 but now it's failing there.  I did notice on FreeBSD that prior to this PR, all create events were handled but that is not the case now.  That being said, I'll chase this down and update the PR.

**Comment by whitlockjc** [comment]
I've cleaned up the sources a little to remove some compiler warnings on non-OSX/Windows.  Progress:

**FreeBSD**

FreeBSD was failing prior to this PR but I wanted to see if there was something I could do about it.  What I've noticed is that the fs event callback `fs_event_cb_dir_multi_file` is called only once even though the test creates 16 files, which should result in 16 separate `fs_event_cb_dir_multi_file` invocations.  `dtruss -e -d -f ./test/run-tests fs_event_watch_dir` shows only 3 `kevent` calls and that doesn't seem right but it seems to match up with what I'm seeing in my debugging.  On a whim, I removed the `EV_ONESHOT` from https://github.com/libuv/libuv/blob/v1.x/src/unix/kqueue.c#L97 and the test passes but I know that is not the fix, and it could very well be a red herring.  I'm working with @indutny on this but I don't know how much of his time I'll have.

**CentOS**

Building an environment as we speak.  No idea why it would stop passing due to these changes but I'll find out soon enough.

**SmartOS**

SmartOS was failing prior to this PR as well and it has similar symptoms to FreeBSD.  I will circle around to SmartOS when I make progress on one of the environments above.

**Comment by whitlockjc** [comment]
I've merged in `v1.x` changes and resolved conflicts.  I'm testing on CentOS right now.

**Comment by whitlockjc** [comment]
I just tested locally on CentOS 7 and `fs_event_watch_dir` passes without issue.  Testing FreeBSD again.

**Comment by whitlockjc** [comment]
I think we're back in the state we were prior to this PR.  FreeBSD and SmartOS are failing as they were before and I think it might be a bigger issue.

@saghul Can you kick off a new CI build to make sure CentOS7 is passing again?

**Comment by saghul** [comment]
Thanks again for the effort you're putting  on this. I'll trigger the CI as
soon as I'm back home.

**Comment by whitlockjc** [comment]
No problem.  If you get a few minutes when you get home, ping me in IRC so we can chat.

**Comment by indutny** [comment]
I have investigated the FreeBSD side. I'm afraid this is just how it works: it coalesces events into one `knote` and returns them at `kevent()` call.

I think we could do following thing - make it create and unlink files in different event-loop ticks. Does it make sense?

**Comment by whitlockjc** [comment]
Yes.  I will update the PR when I've got that approach implemented.  Thanks for your help.

**Comment by whitlockjc** [comment]
The latest version of the PR has `fs_event_watch_dir` passing on FreeBSD, which should allow us to check that one off of #128 if/when this gets merged, and it is also passing on SmartOS now.  *(Both of these were failing prior to this PR.)*

Once @indutny verified that the fs event callback `fs_event_watch_dir` was only being called once and was by design *(expected)*, we realized the only way for this test to pass as written was to create one file per event loop tick and remove one file per event loop tick to make sure that the proper number of fs events were emitted for the created and removed files.

As it stands, this PR not only makes the `fs_event_watch_dir` and `fs_event_watch_dir_recursive` stable, it also fixes `fs_event_watch_dir` on FreeBSD, SmartOS and possibly others where it might had been failing.  Thanks a lot @indutny and @misterdjules.

**Comment by saghul** [comment]
@whitlockjc sorry, yesterday ended up not working for me. CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/133/

**Comment by saghul** [comment]
@whitlockjc can you please rebase again? I pushed a fix for the broken build on Windows.

**Comment by whitlockjc** [comment]
Done.

**Comment by saghul** [comment]
CI: https://jenkins-iojs.nodesource.com/view/libuv/job/libuv+any-pr+multi/135/

**Comment by saghul** [comment]
Great job @whitlockjc! LGTM :+1:

**Comment by whitlockjc** [comment]
Thanks to all that helped: @indutny and @misterdjules  Two excellent mentors.

**Comment by saghul** [comment]
Landed in :sparkles: 8326470 :sparkles: Much obliged, @whitlockjc!

**Comment by whitlockjc** [comment]
WOOHOO!

**Pull request closed without merging**